docs(graphql): Explain the concept of scopes for GraphQL (#24778)
MystenLabs/sui: docs/content/concepts/data-access/graphql-rpc.mdx

- Queries that paginate through history are not available under executed transaction scopes. Before indexing occurs, the system cannot determine where in the history the transaction falls.

### Setting checkpoint scope

You can run queries against a historical checkpoint using `Checkpoint.query`. For [consistent](#consistency) live object set queries (such as fetching an address' owned objects or balances, or an object's dynamic fields), you can continue to paginate using a cursor obtained from a previous query at that checkpoint.

```graphql
query AtCheckpoint($cp: UInt53!) {
  checkpoint(sequenceNumber: $cp) {
    query {
      transactions(last: 5) {
        nodes {
          digest
        }
      }
    }
  }
}

query NextBalancesPage($address: SuiAddress!, $after: String!) {
  address(address: $address) {
    balances(after: $after, first: 10) {
      pageInfo {
        hasNextPage
        endCursor
      }
      nodes {
        objectId
        balance
      }
    }
  }
}
```

> `AtCheckpoint` returns the last 5 transactions to execute as of the end of the checkpoint with sequence number `$cp`, while `NextBalancesPage` will fetch the next 10 balances for `$address` after the page that ended at cursor `$after`. It is also possible to combine both approaches, to ensure that the first page of a live object query is fetched from a specific checkpoint:

Author: ditto

## Scope

GraphQL requests are evaluated in a **scope** that controls the checkpoint being viewed. The GraphQL service will respond to queries as if this is the last checkpoint to have been executed. By default this is set to the latest checkpoint that the service has all data for. To avoid returning partial responses, the service will not allow queries to specify a later checkpoint than this, but it can be set to an earlier checkpoint to perform historical queries.

The scope optionally provides a **root object** bound which only applies to queries that fetch dynamic fields. This fetches the dynamic fields as they were at the end of a specific checkpoint or when their root object was at some given version. For any wrapped or child (object-owned) object, its root object is defined recursively as:

- the root object of the object it is wrapped in, if it is wrapped,
- the root object of its owner, if it is owned by another object, otherwise

> - the object itself.

Reviewer: ```suggestion
- The root object of the object it is wrapped in, if it is wrapped
- The root object of its owner, if it is owned by another object
- The object itself
```

If a dynamic field's root object has version `v`, its own version, `w` is the latest version such that `w <= v` -- that is, the latest version of the dynamic field that existed when its root object was at version `v`.

If a root object bound is not provided, dynamic fields are fetched at the checkpoint being viewed, while if a checkpoint-based root object bound exists, it does not impact the checkpoint being viewed.

Finally, the GraphQL service treats queries nested under executed (`Mutation.executeTransaction`) and simulated (`Query.simulateTransaction`) transactions as being in a special scope that exists just after the transaction that was executed or simulated, without having indexed that transaction.

The scope a query is evaluated in impacts which fields are available. In particular:

- live object set queries are not available under executed transaction scopes, or if there is a root object bound given in terms of a specific version, because these queries rely on data that is indexed at the checkpoint level.

> - queries that paginate through history are not available under executed transaction scopes, because it's not possible ot pin-point where in the history the transaction will fall before it has been indexed.

Reviewer: ```suggestion
- Queries that paginate through history are not available under executed transaction scopes. Before indexing occurs, the system cannot determine where in the history the transaction falls.
```

        endCursor
      }
      nodes {
        objectId
        balance
      }
    }
  }
}
```

`AtCheckpoint` will return the last 5 transactions to execute as of the end of the checkpoint with sequence number `$cp`, while `NextBalancesPage` will fetch the next 10 balances for `$address` after the page that ended at cursor `$after`. It is also possible to combine both approaches, to ensure that the first page of a live object query is fetched from a specific checkpoint:

```graphql
query FirstBalancesPageAtCheckpoint($address: SuiAddress!, $cp: UInt53) {
  checkpoint(sequenceNumber: $cp) {
    query {
      address(address: $address) {
        balances(first: 10) {
          pageInfo {
            hasNextPage
            endCursor
          }
          nodes {
            objectId
            balance
          }
        }
      }
    }
  }
}
```

Responses to these queries are subject to [retention](#retention). If `$cp` is outside the retention window `Query.transactions`, `AtCheckpoint` will return no results, while an attempt to continue paginating `NextBalancesPage` at a checkpoint outside the consistent range will return an error.

### Setting Root Version Scope

> Queries nested under the fetch of an object at a specific version are scoped by a root object bound at its version. For example in the following query, the dynamic field with name `42u64` is fetched as it existed when its owning object was at version `$v`:

Reviewer: ```suggestion
Queries nested under the fetch of an object at a specific version are scoped by a root object bound at its version. For example, in the following query, the dynamic field with name `42u64` is fetched as it existed when its owning object was at version `$v`:
```

```graphql
query NestedDynamicFields($id: SuiAddress!, $r: UInt53!) {
  object(address: $id, rootVersion: $v) {
    asMoveObject {
      asDynamicField {
        value {
          ... on MoveValue {
            extract(path: "foo->[42u64]") {
              json
            }
          }
        }
      }
    }
  }
}
```

Queries that return an `Address` can be used to make nested dynamic field queries on wrapped objects. An `Address` does not have an associated version, but `Query.address` can accept `rootVersion` or `atCheckpoint` parameters to set root object bounds for nested dynamic field queries. The following query fetches a dynamic field with name `42u64` owned by an object with address `$id` -- which can belong to a wrapped object -- as it was when its root object was at version `$r`:

```graphql
query WrappedDynamicField($id: SuiAddress!, $r: UInt53!) {
  address(address: $id, rootVersion: $r) {
    dynamicField(name: { literal: "42u64" }) {
      value {
        ... on MoveValue {
          json
        }
      }
    }
  }
}
```

> While nested within a scope that has a root object bound, it is possible to reset or override the bound, using the `objectAt` or `addressAt` queries. These fields query the same entity but at a different position in its history. If these fields are not provided any parameters, they fetch the state of the object at the latest known checkpoint to the GraphQL service. The following query fetches the latest version of an object that was previously fetched at version `$v`:

Reviewer: ```suggestion
While nested within a scope that has a root object bound, you can reset or override the bound using the `objectAt` or `addressAt` queries. These fields query the same entity but at a different position in its history. If these fields are not provided any parameters, they fetch the state of the object at the latest known checkpoint to the GraphQL service. The following query fetches the latest version of an object that was previously fetched at version `$v`:
```
